web/src/games/cardtable/game.ts

```diff
@@ -0,0 +1,194 @@
+    //this uses Immer mutability under the hood
+    let source: ICard[] = getNamedContainer(G, cm.from.container);
+    let target: ICard[] = getNamedContainer(G, cm.to.container);
+    let moving: ICard[] = source.splice(cm.from.ordinal, cm.from.cardcount ? cm.from.cardcount : 1);
+    target.splice(cm.to.ordinal, 0, ...moving);
+  } else return INVALID_MOVE;
+};
+
+const cutDeck = (G: IG, ctx: Ctx, depth: number) => {
+  // by convention top card is at 0
+  //depth can never be greater than G.deck.length
+  //normalize to a value < G.deck.length
+  var ndepth = depth % G.deck.length;
+
+  let top: ICard[] = ndepth > 0 ? G.deck.slice(0, ndepth) : G.deck.slice(0, G.deck.length + ndepth);
+  let bottom: ICard[] = ndepth > 0 ? G.deck.slice(ndepth, G.deck.length) : G.deck.slice(ndepth);
+  top.unshift(...bottom);
+  return { ...G, deck: top };
+};
+
+export const moves = {
+  cutDeck,
+
+  moveCards,
+
+  deal(G: IG, ctx: Ctx) {
+    let fresh: ICard[] = dealCribbage.fresh.slice(0);
+    let { pattern } = dealCribbage;
+    //player 0 sits north, player 1 sits south
+    let dealer = ctx.playerID;
+    fresh = ctx.random.Shuffle(fresh);
+    let draftNorth = [];
+    let draftSouth = [];
+    let holders = dealer === '1' ? [draftNorth, draftSouth] : [draftSouth, draftNorth];
+    for (var i = 0; i < pattern.hand; i++) {
+      for (var j = 0; j < pattern.players; j++) {
+        holders[j % pattern.players].push(fresh.pop());
+      }
+    }
+    G.hands.north.melds.length = 0;
+    G.hands.north.played.length = 0;
+    G.hands.north.private.length = 0;
+    G.hands.north.tricks.length = 0;
+    G.hands.north.tricks.length = 0;
+    G.hands.north.held.length = 0;
+    G.hands.north.held = draftNorth;
+    G.hands.east.private.length = 0; //used for crib
+    G.hands.south.melds.length = 0;
+    G.hands.south.played.length = 0;
+    G.hands.south.private.length = 0;
+    G.hands.south.tricks.length = 0;
+    G.hands.south.held.length = 0;
+    G.hands.south.held = draftSouth;
+    G.stock = [];
+    G.deck = fresh;
+  },
+
+  putToCrib(G: IG, ctx: Ctx, idx: number) {
+    let who = ctx.playerID;
+    let playFrom: ILocation = findFromForPlayer(G, ctx, idx, who);
+    let playTo: ILocation = findCribForPlayer();
+    let cribMove = {
+      from: playFrom,
+      to: playTo,
+    };
+    moveCards(G, ctx, cribMove);
+  },
+
+  play(G: IG, ctx: Ctx, idx: number) {
+    let who = ctx.playerID;
+    let playFrom: ILocation = findFromForPlayer(G, ctx, idx, who);
+    let playTo: ILocation = findPlayedForPlayer(G, ctx, who);
+    let playMove = {
+      from: playFrom,
+      to: playTo,
+    };
+    moveCards(G, ctx, playMove);
+  },
+  cutShowTurn(G: IG, ctx: Ctx, idx: number) {
+    cutDeck(G, ctx, idx);
+    G.deck.length = 1;
+    return;
+  },
+
+  flipCrib(G: IG) {
+    G.hands.east.cribFlipped = !G.hands.east.cribFlipped;
+  },
+};
+
+const getNamedContainer = (G: IG, name: String) => {
+  let path: string[] = name.split('.');
```

Comment:
Use an array instead of splitting a string. It is a bad practice to use a string to store structured data.

Reply:
So, since I'm still learning my chops with JS you're going to see a lot of mealy code--be patient and I'll be open to moving my solution so it looks like what you're used to.  I'm not sure that:

keysPath = ['mycontainer', 'below','cards','blender'] is more understandable than 'mycontainer.below.cards.blender' but I'll move this into my next release as an issue.  I was asking you folks for guidance as much as defending my code line.  I'm happy with the progress and am getting comfortable with the data model -- this app may not fully represent your standards, but it would make a nice tutorial -- the state model and ui are very simple which makes a good starter and the abstractions are very familiar to most of us -- table, deck, hand, deal, cut, all of them built around just a couple of very sturdy primitive mutators -- if this core is moved to common my game is about 200 lines of JS.  I'm going to eventually have more Jest/Enzyme coverage as the UI matures, but the game has very few lines that aren't covered.

Reply:
using keysPath = ['mycontainer', 'below','cards','blender']  isnt storing a deep object, just a reference in a data structure that actually means "several strings" instead of one . I wont block you on this, but it is a bad practice.

Reply:
No updates as of yet.  Still tracking this request 

> Use an array instead of splitting a string. It is a bad practice to use a string to store structured data.

Can you explain what you mean by use an array rather than splitting?  I'm going to take a long look at the game after it deploys latest changes.  I'm willing to implement this function any which way we deem.  The other factor I'm considering is the request to pull standard components to common -- I think that's a great idea and it would save a lot of work -- work on this product in advance of that moment should be minimal.  If it's good enough for WIP deployment i would appreciate the lift from the community at refining the approaches as we see fit.